File: 장세훈/LeetCode/easy/LinkedListCycle.java
class LinkedListCycle {
    /*
        141. Linked List Cycle
        https://leetcode.com/problems/linked-list-cycle/description/?envType=study-plan-v2&envId=top-interview-150

        링크드 리스트가 순환하는지 판별하는 문제.
        노드를 순회하며 set에 넣음
        set에 존재하면 true 반환

     */
    public boolean hasCycle(ListNode head) {
        Set<ListNode> set = new HashSet<>();

        while (head != null) {
            set.add(head);
            head = head.next;

            if (set.contains(head))
                return true;
        }

        return false;
    }
}

File: 정수빈/LeetCode/easy/linked-list-cycle.py
# https://leetcode.com/problems/linked-list-cycle/?envType=study-plan-v2&envId=top-interview-150

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:

        visited = set()
        node = head
        while node:
            if node.next == None:
                return False
            elif node in visited:
                return True
            visited.add(node)
            node = node.next
        return False
            
            
